### PR TITLE
rgw/s3-notifications: use user-name/password topic attributes for SASL authentication

### PR DESCRIPTION
**Pull request by igomon-bloomberg**
rgw/s3-notifications: use user-name/password topic attributes for SASL authentication.
RGW allows to specify user-name:password as part of the push_endpoint URI. This does not work however when the user-name/password themselves contain : (colon) symbol. This PR implements an alternative way to provide those via topic attributes.

Fixes: https://tracker.ceph.com/issues/65050
Signed-off-by: Igor Gomon ([igomon@bloomberg.net](mailto:igomon@bloomberg.net))

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by yuvalif** [comment]
@igomon-bloomberg fix looks good. why did you clode the PR?
2 comments if you decide to reopen:
* please make sure these attributes are only sent over secure connection (or if the workaround conf is used...): https://github.com/igomon-bloomberg/ceph/blob/00aa1b2302663851f1259b279858a7f5664ccf4d/src/rgw/rgw_rest_pubsub.cc#L38
* please add documentation: https://github.com/ceph/ceph/blob/main/doc/radosgw/notifications.rst

**Comment by igomon-bloomberg** [comment]
> @igomon-bloomberg fix looks good. why did you clode the PR? 2 comments if you decide to reopen:
> 
> * please make sure these attributes are only sent over secure connection (or if the workaround conf is used...): https://github.com/igomon-bloomberg/ceph/blob/00aa1b2302663851f1259b279858a7f5664ccf4d/src/rgw/rgw_rest_pubsub.cc#L38

@yuvalif Just to be clear: are you saying that the same validation logic applies to the user name/password provided via topic attributes? 

> * please add documentation: https://github.com/ceph/ceph/blob/main/doc/radosgw/notifications.rst
Sure.

**Comment by igomon-bloomberg** [comment]
> it does not really matter if user/password are sent in the URL or in the http header. in both cases, they are sent in cleartext, which causes a security issue. so, we want to make sure that they are only sent if TLS is used (or the admin specified that it is allowed to send them in cleartext using the global conf parameter)

@yuvalif there is already a check for sending the username/password over unencrypted connection here:
https://github.com/igomon-bloomberg/ceph/blob/00aa1b2302663851f1259b279858a7f5664ccf4d/src/rgw/rgw_kafka.cc#L568
That check is preceded by new logic that overrides URL-provided user-name/password with topic attributes (if any). This verification, though, happens when attempting to connect to the endpoint. Is it not sufficient?

**Comment by igomon-bloomberg** [comment]
@yuvalif I have added REST-level checks in the latest commit

**Comment by yuvalif** [comment]
* please cleanup the commits. you just need your 2 commits - please remiove the rest
* please update docs

File: src/rgw/rgw_rest_pubsub.cc [hunk 1]
```diff
@@ -45,11 +45,31 @@ bool validate_and_update_endpoint_secret(rgw_pubsub_dest& dest, CephContext *cct
...
+
+  const auto& args=ri.args;
+  auto topic_user_name=args.get_optional("user-name");
+  auto topic_password=args.get_optional("password");
+
+  // check if username/password was already supplied via topic attributes
+  // and if also provided as part of the endpoint URL issue a warning
+  if (topic_user_name.has_value()) {
+    if (!user.empty()) {
+      ldout(cct, 0) << "WARNING: username provided via both topic attributes and endpoint URL: using topic attributes" << dendl;
```
[the remark this answers is not in the document]
Comment (igomon-bloomberg) [reply src/rgw/rgw_rest_pubsub.cc:57]: Sure, though I was just following [this](https://github.com/igomon-bloomberg/ceph/blob/9d51002066f3a31408c764ee9a81de340c0cc638/src/rgw/rgw_rest_pubsub.cc#L29)

**Review by yuvalif — APPROVED** [review]

**Comment by igomon-bloomberg** [comment]
@yuvalif @cbodley I just pushed another commit. This is to prevent topic's password from being dumped to RGW logs.

**Comment by yuvalif** [comment]
> @yuvalif @cbodley I just pushed another commit. This is to prevent topic's password from being dumped to RGW logs.

but this would prevent from logging any post command content.
is there a way to only prevent logging of topic creations?

**Comment by igomon-bloomberg** [comment]
Right. Well, @cbodley suggested that line was unnecessary and no harm would be done If I just removed it. My other suggestion was to just strip the post body of password attribute and dump the rest. Or do you prefer I find a way to prevent logging on topic creation only?

**Comment by yuvalif** [comment]
> Right. Well, @cbodley suggested that line was unnecessary and no harm would be done If I just removed it. My other suggestion was to just strip the post body of password attribute and dump the rest. Or do you prefer I find a way to prevent logging on topic creation only?

if this log is not useful, juts remove it :-)

**Comment by igomon-bloomberg** [comment]
That's what the last commit does - it just removes that line :-)

**Comment by igomon-bloomberg** [comment]
@yuvalif @cbodley I see that the doc review has been pending for a while, do you think this PR will make it into Squid?

**Comment by cbodley** [comment]
> @yuvalif @cbodley I see that the doc review has been pending for a while, do you think this PR will make it into Squid?

doc approval isn't required for merge. Eric's picked up this for qa so there should be plenty of time to get it to squid

**Comment by ivancich** [comment]
Teuthology results:
* https://pulpito.ceph.com/ivancich-2024-05-08_17:43:09-rgw-wip-eric-testing-1-distro-default-smithi/
* https://pulpito.ceph.com/ivancich-2024-05-08_17:43:43-rgw-wip-eric-testing-1-distro-default-smithi/

**Pull request merged**